chore: Remove unsafe use of from_raw_parts in Parquet decoder

andygrove opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Part of https://github.com/apache/datafusion-comet/issues/507

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

The documentation for `from_raw_parts` states that the the memory must be "properly aligned" and Rust 1.78 added debug assertions that fail if the memory is not properly aligned. We are seeing those assertions fail when we run with Rust 1.78.

https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

Refactor `make_int_variant_impl` to follow the same pattern as other decode methods and use `read_unaligned` and `write_unaligned` instead of incorrect use of `from_raw_parts`.

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Existing tests.

andygrove commented on core/src/parquet/read/values.rs:
```diff
@@ -455,23 +455,21 @@ macro_rules! make_int_variant_impl {
 
                 while num - i >= 32 {
                     unsafe {
-                        let in_slice = std::slice::from_raw_parts(in_ptr, 32);
```
This is unsafe because there is no guarantee that the buffer is properly aligned for u32 typw (in_ptr is defined earier as `let mut in_ptr = &src_data[src_offset..] as *const [u8] as *const u8 as *const u32;`).

parthchandra commented on core/src/parquet/read/values.rs:
```diff
@@ -443,49 +443,21 @@ macro_rules! make_int_variant_impl {
-                let mut i = 0;
-                let mut in_ptr = &src_data[src_offset..] as *const [u8] as *const u8 as *const u32;
-
-                while num - i >= 32 {
```
So this loop is effectively not needed?

andygrove replied:
Yes, I combined the two loops into one. I explained more in https://github.com/apache/datafusion-comet/pull/549#issuecomment-2159574852

parthchandra reviewed (commented):
Should we investigate why we have unaligned memory in the first place?

andygrove commented:
> Should we investigate why we have unaligned memory in the first place?

The input is a slice of `u8`, so it has no particular alignment, and we are trying to turn it into a slice of `u32` using `from_raw_parts`. When we run this with Rust 1.78 we get this error:

```
unsafe precondition(s) violated: slice::from_raw_parts requires the pointer to be aligned and non-null, and the total size of the slice not to exceed `isize::MAX`
```

The original code seems to be copying data over using two loops. The first loop was copying in slices of 32 items and then the second loop copied any remaining items. I can only assume that this was some sort of optimization, but it relied on undefined behavior, which is now caught by Rust 1.78.

I simplified the code to use a single loop and removed the unsafe optimization. The code is now almost identical to the decode method for other types such as this one, which I based it on:

```rust
impl PlainDecoding for Int32To64Type {
    fn decode(src: &mut PlainDecoderInner, dst: &mut ParquetMutableVector, num: usize) {
        let src_ptr = src.data.as_ptr() as *const i32;
        let dst_ptr = dst.value_buffer.as_mut_ptr() as *mut i64;
        unsafe {
            for i in 0..num {
                dst_ptr
                    .add(dst.num_values + i)
                    .write_unaligned(src_ptr.add(src.offset + i).read_unaligned() as i64);
            }
        }
        src.offset += 4 * num;
    }
```

I don't know why we were trying to do the optimization for some cases and not others. Perhaps @sunchao can remember.

sunchao reviewed (approved):
This looks correct. I also forgot exactly why the code is written in the original way: maybe it is for some sort of loop unrolling. You could check the internal `git blame` results to see if there is some hint.

andygrove commented:
Thanks for the review @sunchao!

parthchandra commented:
> The input is a slice of `u8`, so it has no particular alignment, and we are trying to turn it into a slice of `u32` using `from_raw_parts`.

What I meant was that we expect the source buffer to be aligned on a word boundary. Subsequently, all offsets applied to the starting address are at least a multiple of size of word so every input should be aligned.
It appears at least one of my assertions above is not true and it might be useful to know why.
Unaligned access has two potential problems - on some platforms, it may cause a SIGBUS and, more importantly, access to an unaligned address is usually much slower.
